ports/stm32f4/common-hal/busio/UART.c

```diff
@@ -36,45 +36,455 @@
+
+    //start the interrupt series
+    if ((HAL_UART_GetState(&self->handle) & HAL_UART_STATE_BUSY_RX) == HAL_UART_STATE_BUSY_RX) {
+        mp_raise_ValueError(translate("Could not start interrupt, RX busy"));
+    }
+
+    //start the recieve interrupt chain
+    HAL_NVIC_DisableIRQ(self->irq); //prevent handle lock contention
+    HAL_UART_Receive_IT(&self->handle, &self->rx_char, 1);
+    HAL_NVIC_SetPriority(self->irq, UART_IRQPRI, UART_IRQSUB_PRI);
+    HAL_NVIC_EnableIRQ(self->irq);
+
+    errflag = HAL_OK;
 }
 
 bool common_hal_busio_uart_deinited(busio_uart_obj_t *self) {
-    return 0;
+    return self->tx->pin == mp_const_none;
 }
 
 void common_hal_busio_uart_deinit(busio_uart_obj_t *self) {
+    reset_pin_number(self->tx->pin->port,self->tx->pin->number);
+    reset_pin_number(self->rx->pin->port,self->rx->pin->number);
+    self->tx = mp_const_none;
+    self->rx = mp_const_none;
+    gc_free(self->rbuf.buf);
+    self->rbuf.size = 0;
+    self->rbuf.iput = self->rbuf.iget = 0;
 }
 
-// Read characters.
 size_t common_hal_busio_uart_read(busio_uart_obj_t *self, uint8_t *data, size_t len, int *errcode) {
-    return 0;
+    if (self->rx == NULL) {
+        mp_raise_ValueError(translate("No RX pin"));
+    }
+
+    size_t rx_bytes = 0;
+    uint64_t start_ticks = ticks_ms;
+
+    // Wait for all bytes received or timeout, same as nrf
+    while ( (ringbuf_count(&self->rbuf) < len) && (ticks_ms - start_ticks < self->timeout_ms) ) {
+        RUN_BACKGROUND_TASKS;
+        //restart if it failed in the callback
+        if(errflag != HAL_OK) {
+            errflag = HAL_UART_Receive_IT(&self->handle, &self->rx_char, 1);
+        }
+        // Allow user to break out of a timeout with a KeyboardInterrupt.
+        if ( mp_hal_is_interrupted() ) {
+            return 0;
+        }
+    }
+
+    // Halt reception
+    HAL_NVIC_DisableIRQ(self->irq);
+    // copy received data
+    rx_bytes = ringbuf_count(&self->rbuf);
+    //Used for debugging
+    //mp_printf(&mp_plat_print, "Read: count:%d, buffer location%p, er:%d\n", rx_bytes, &self->rbuf,errflag);
+    rx_bytes = MIN(rx_bytes, len);
+    for ( uint16_t i = 0; i < rx_bytes; i++ ) {
+        data[i] = ringbuf_get(&self->rbuf);
+    }
+    HAL_NVIC_EnableIRQ(self->irq);
+
+    if (rx_bytes == 0) {
+        *errcode = EAGAIN;
+        return MP_STREAM_ERROR;
+    }
+    return rx_bytes; 
 }
 
 // Write characters.
 size_t common_hal_busio_uart_write(busio_uart_obj_t *self, const uint8_t *data, size_t len, int *errcode) {
-    return 0;
+    if (self->tx == NULL) {
+        mp_raise_ValueError(translate("No TX pin"));
+    }
+
+    HAL_NVIC_DisableIRQ(self->irq);
+
+    if (HAL_UART_Transmit(&self->handle, (uint8_t *)data, len, self->timeout_ms) != HAL_OK) {
+        mp_raise_ValueError(translate("UART write error"));
+    }
+
+    HAL_UART_Receive_IT(&self->handle, &self->rx_char, 1);
+    HAL_NVIC_SetPriority(self->irq, UART_IRQPRI, UART_IRQSUB_PRI);
```

Comment:
Why do you set the priority here? Should it be set once in the constructor or in uart_reset or uart_init or something?